trie/database.go

```diff
@@ -751,39 +765,62 @@ func (db *Database) commit(hash common.Hash, batch ethdb.Batch) error {
 		if err := batch.Write(); err != nil {
 			return err
 		}
-		batch.Reset()
+		db.lock.RUnlock()
+		{
+			db.lock.Lock()
+			batch.Replay(uncacher)
+			db.lock.Unlock()
+			batch.Reset()
+		}
+		db.lock.RLock()
```

Comment:
Yes, this is a cluster-f*ck. That said, the trie database write access is serialized by the blockchain importer, since we can't import two blocks concurrently. The only thing this code needs to ensure is that API CALL operations, sync state retrievals, etc don't mess up. The blockchain guarantees that no write will intervene here.